### DISCUSSION TITLE
Does Windows App SDK need better support for periodic background tasks?

### DISCUSSION THREAD

**Discussion by JosHuybrighs** [microsoft/WindowsAppSDK#2322] [Q&A]
I am porting my SyncFolder UWP/Desktop-Bridge app to a WinUI 3 - Desktop app (using Windows App SDK 1.0 and a package/wap project).
My app allows users to configure sync/backup tasks for selected folders and target devices and allows users to automatically schedule these tasks using a configured time schedule.
In the current UWP/DesktopBridge version the scheduling occurs with a UWP timer-triggered background task, which every 15 minutes checks whether a task is due and if so launches a win32 project to do the actual copying/syncing.
The win32 process uses a `AppService Connection` to communicate status and progress with the UI user-app in case the user would have opened the app.
With WinUI3-Desktop I can finally get rid of the unreliable `AppService Connection` because it allows me to have a single app/process that can both show the UI when necessary or run hidden and execute the sync/backup tasks when triggered in the background.

The challenge with Windows App SDK now is: which concept to use for running a periodic task?
There seems to be 3 possible approaches, but none of them offers enough capabilities.

1. **Timer-triggered background task**, as it is done in UWP.
Today this requires an OOP background task and, if you follow the description I gave in an earlier discussions topic [#2314](https://github.com/microsoft/WindowsAppSDK/discussions/2314), is not so difficult to do.
The major problem however with this concept is that such a task doesn't run while the screen is in lock state. That is not ideal for the kind of app I am publishing.

2. **A startup-task that periodically executes code**.
Such task would be added to the package and will do whatever it needs to do (check if sync/copy tasks are due in my case), wait 15 minutes, check again. It will keep on doing this until the task gets killed.
This is relatively simple to implement using `Application.Run(context)` in a windows forms project (without an active window) and a `Timer` that fires every 15 minutes. It also seems to allow that execution keeps on occurring while in the lock screen.
2 problems however with this:

- The process uses computer resources permanently.
- When updating the app the deployment/provisioning service in Windows kills the startup-task, which generates an error in the windows event log saying that the process hangs. I tried to hook up the WM_CLOSE message but it look like this message isn't dispatched when the app gets killed by the Windows deployment service.

3. **Register the app with the Windows Task Schedular**.
That seems to be the ideal solution since it allows execution to happen in all kind of circumstances, like in lock screen, force to go out of sleep state, etc.
The major issue however with this is that there is no possibility to deregister the app from the Task Schedular when the app is uninstalled.
Windows App SDK supports a UpdateTask (to support app updates) but there is no "UninstallTask".
Maybe of interest: there is an issues topic [#16](https://github.com/microsoft/WindowsAppSDK/issues/16) that discusses the need to support 'uninstall' in the MSIX manifest and support for a Uninstall-task. 

I am interested to learn how other people would approach this. Support for a "UninstallTask" seems to be a good solution for everyone who doesn't want to use a seperate installer product.

**Comment by jonwis** [comment]
Great questions! I'd normally say to use the Timer-triggered BG task.

Hey @andreww-msft  and @aeloros - seems like a great "make background tasks easier" space.  Do you have a good example of wiring up a bg task that drives the activation into a running app, like the other AppLifecycle work you've done?

**Reply by JosHuybrighs** [reply]
@jonwis , @DrusTheAxe 

It is almost 2 years ago since I brought up the need for better support for background tasks in WinUI/WinAppSDK.
The least I can say is that the situation is getting worse.

The only solution proposed by Microsoft to have some kind of timer-triggered BG task handling in a C# WinUI3/WinAppSDK desktop app/package is to register an out of process BackgroundTask. InProc BG tasks are not possible in C#.
Today, this is simply not possible for the following reasons:

- The task needs a WinRT Runtime Component. If you write code in the task that accesses today's API's like StorageFolder, ApplicationDataContainer, etc. the BG task doesn't run. (is interrupted after launch).
- If you include a shared project library in the task project (what any sensible person would do to avoid duplicated code) VS 18.3 refuses to create the package. See [Visual Studio Issue](https://developercommunity.visualstudio.com/t/Cant-build-WinUIWinAppSDK-desktop-app/10525478?port=1025&fsid=2cf7a557-a22c-45cb-bf88-41397708e1d5&ref=native&refTime=1703335036637&refUserId=0bad7782-f8c8-49f0-8014-b1c221e00c58).

One possible way to overcome the above is to have a BG task that doesn't do any of the above and simply launches the main app, which then must carry out the 'background' request without having a MainWindow.
Is this what I should do? Launch the app every 15 minutes? Serious?

I recently opened discussion [4014](https://github.com/microsoft/WindowsAppSDK/discussions/4014) about the same thing. The only response I got was that BG tasks are not part of the WinAppSDK. OK, but to be honest I am not interested whether this is a WinAppSDK, or a WinUI3 thing. I simply need something that allows me to use C# and supports a background task that is triggered by a timer and has access to important app model data.

I can only see 2 solutions:

1. Allow InProc background tasks. This way all of the app's model and settings properties are accessible from within the task.
2. Provide a package Uninstall trigger to allow WinAppSDK Desktop apps that would use alternative ways to deal with timer-triggered BG tasks, e.g. by using the Windows Task Schedular. That trigger can then be used to deregister things that are not automatically cleaned up after uninstall. Examples: 1 or more Task Schedular registrations, a Firewall rule that has been created, etc. 

To be honest, I prefer the last solution with an Uninstall trigger. It offers most flexibility. I don't think that without a Uninstall trigger a WinUI Desktop App will ever be able to cover all possible use cases and replace 'old' win32, winforms, wpf type of solutions.

**Reply by DarranRowe** [reply]
As a completely independent opinion.

Honestly, I don't see any real resolution to this without Windows itself being updated to support it. Since the Windows App SDK is not the Windows SDK, then I would say that there is very little to be done on this side. All of the important work would end up needing to be done in Windows.

The samples for in process background tasks use CoreApplication, in particular, the EnteredBackground and LeavingBackground events. CoreApplication, besides the Properties property, is not available for any desktop application to use on any of the current release versions of Windows. This matters because WinUI 3 project types are desktop applications.
Windows would have to allow a desktop application to use the CoreApplication functionality, in particular the app lifecycle functionality. Desktop applications, even when packaged using the Windows packaging technology, do not automatically get app lifecycle functionality. There is a little more to add here, but I'll go into more details soon.

For the uninstall trigger, again, this would need Windows level changes since Windows is what controls package installation and removal. But this also seems to go against some of the core ideas of application packages. That is the clean installation and removal. The firewall rules as an example should cleanly uninstall if you are using [desktop2:FirewallRules](https://learn.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-desktop2-firewallrules). The most ideal thing here, especially since Windows would have to be updated, would be some package extension to allow more registrations.

The addition to the app lifecycle functionality is interesting. If you look at the manifest schema for [Application](https://learn.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-application#attributes-and-elements), under the attributes, there is one named desktop11:AppLifecycleBehavior. It is important to point out that the release versions of Windows supports at most desktop8 right now. However, if you look at the insider Windows SDK then there are some really interesting artefacts.

![Screenshot 2023-12-26 112048](https://github.com/microsoft/WindowsAppSDK/assets/52577874/ca197a73-5e50-43c3-997d-8a8908ef1563)

Yes, the Windows Insider SDK version 26016 has a DesktopManifestSchema_v11.xsd as part of the package schema validation files. This version of Windows, 26016, is the latest version available in the Canary insider channel. This is interesting for two reasons. The documentation for desktop11:AppLifecycleBehavior states that a packaged desktop application is able to opt into the UWP app lifecycle. Secondly, after prodding through this version of the Windows SDK a little, there is nothing added for desktop applications specifically to use the equivalent of CoreApplication's lifetime events. While I have yet to test it, this seems to imply that setting desktop11:AppLifecycleBehavior to systemManaged, a desktop application running inside an application package can use the CoreApplication lifecycle events. We'll find out more when this functionality is enabled in one of the other insider channels, but this should allow for in process background tasks.

**Reply by JosHuybrighs** [reply]
@DarranRowe 
Thanks for explaining this. Looks like you have quite some insights in what is going on with these new API's. Looks like with a WinUI3 package there are indeed constraints one has to live with. Some will remain, some (like hopefully inproc background tasks) might be resolved at some point in time.

The uninstall trigger I mentioned came to my mind more out of frustration of those constraints. But, I have to admit that I do like how in a WinUI3 package one doesn't need to worry about cleanup of registrations. Good point about the desktop2:FirewallRules extension: I didn't know it exists and so I will definitely will use that one.

Let's hope that c# inproc background tasks soon will become possible.
In the mean time I will restrict the code in the background task to simply launch the main exe process. Not ideally since it runs every 15 minutes but I don't have the feeling that it is a big resources consumer.

**Reply by DrusTheAxe** [reply]
>Support for a "UninstallTask"

@JosHuybrighs Why are you interested? Your post is about periodic background tasks and while I can imagine why a 'SyncFolder' app would find such a thing useful, I don't see mentioned how or why an 'uninstall task' would be helpful or used. Could you elaborate?

**Reply by DrusTheAxe** [reply]
>But this also seems to go against some of the core ideas of application packages. That is the clean installation and removal. 

@DarranRowe Indeed. MSIX's deployment engine (i.e. Windows) strives to provide clean and reliable servicing (install and remove, as well as update, reset and so on). Ideally any altered system state you need can be _declarative_ via snippets in `appxmanifest.xml` or, if you party on the system at runtime, you party on resources and locations that are manageable by MSIX and cleanly serviced, e.g. ApplicationData, writing to what _you_ see as HKCU but i/o virtualization prevents from impacting others and Windows can cleanly remove if/when the time comes.

But...Windows is a big system, and while MSIX is and can be expansive, there are limits. You can disable MSIX savvy in several ways (e.g. &lt;cough&gt;[Flexible Virtualization](https://learn.microsoft.com/windows/msix/desktop/flexible-virtualization)&lt;/&gt;). MediumIL processes can further party on resources not managed by MSIX. Packaged desktop apps can even run elevated and party on even more unmanaged (and unmanageable) resources across the system.

There are limits to MSIX's super powers. This is why windows.preinstallConfigTask and windows.updateTask were invented, and why [uninstall/reset/repair equivalents have been proposed](https://github.com/microsoft/WindowsAppSDK/issues/2779).

>The most ideal thing here, especially since Windows would have to be updated, would be some package extension to allow more registrations.

Absolutely. We routinely expand registration options. The question is, what functionality would you like to see added? What pain points do you face that could be aided by richer MSIX functionality? What enhancements would boost your development experience, reliability, performance or other quality of life improvements? Please share your feedback. It's a large and diverse community and ecosystem so input is always welcome.

P.S. We don't always need to update Windows to [expand MSIX functionality](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/preview-channel#msix-packaging). Devil's in the details of course but it's good to have a broad palette of options to paint with :-)

**Reply by JosHuybrighs** [reply]
@DrusTheAxe

> Why are you interested? Your post is about periodic background tasks and while I can imagine why a 'SyncFolder' app would find such a thing useful, I don't see mentioned how or why an 'uninstall task' would be helpful or used. Could you elaborate?

The main reason why the Uninstall trigger could be usefull is the current minimum of 15 minutes for periodic background tasks. SyncFolder is a file/folder sync/backup tool that executes tasks periodically. Some users have expressed the wish/need? to have faster scheduling, like every 2 minutes. I am not sure whether I actually want to offer this but if I would then the only way that comes to my mind is to use the Windows Task Scheduler and register all the tasks there that need the faster scheduling. That will require some kind of uninstall trigger to remove the registrations. Allowing faster periodic background tasks would of course be a much better solution.

Another reason which can even be more debated, is that some users, luckily enough only a few, asked for functionality where the tool more or less on a permanent base looks at designated files to see if they are modified and then carries out the sync/backup. That probably needs some kind of Windows Service. I tend to think that these kind of services however are beyond the scope of my app.

The above 2 constraints are not really deal-breakers for the next major update of the app (where UWP-DesktopBridge is migrated to WinUI3/WinAppSDK). The only annoying thing is the lack of InProc background tasks in C# and the impossibility in the OOP background task to use a shared library for model and settings data. In the WinRT runtime component one cannot access StorageFolder, ApplicationDataContainer, etc.
As I mentioned before, I solved this by don't doing anything except launching the "main" app every 15 minutes. The app then runs without a MainWindow and carries out what actually the background task would do if the above constraints wouldn't exist, like checking if a sync/backup task is due.

One curious thing about this: It looks like when Windows is in sleep mode the periodic task keeps on being fired but the launch of the app seems to get queued. The side effect of this: when Windows goes out of sleep mode suddenly the app is launched 10 to 20 times.

**Reply by DarranRowe** [reply]
@JosHuybrighs 
One question that I have been meaning to ask. Since [Mapping UWP features to the Windows App SDK](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/migrate-to-windows-app-sdk/feature-mapping-table) has some interesting implications. First, C++ has no such thing as in process background tasks. Secondly, all in process background tasks need to be migrated to COM background tasks, which seriously implies C# too. Have you actually tried using this method for your background tasks?
The reason why I ask this is that, even though this COM server is still out of process, it runs with the same integrity level as the application. Extensions like this are also activated using the activate as package activation method, basically it means that the COM server will have the same identity as the application itself.

**Reply by DrusTheAxe** [reply]
> C++ has no such thing as in process background tasks

They're called threads 😋

Historically standard C++ had no formal 'task' API like .NET. Recent iterations have added std::packaged_task, std::async and of course co_await/yield/return. I don't know if you'd consider these equivalent.

They're not language independent but they already have rich language integration and of course established market presence.

Inproc execution poses additional complications (different than OOP, but also more challenging in some ways) and many languages already have mechanisms so it's always a question of is this something Windows could provide sufficient value add? Even if we built a better mousetrap, how much awesomely better could it be? And how much traction would it get given the established alternatives? Tough ROI.

**Reply by DrusTheAxe** [reply]
> all in process background tasks need to be migrated to COM background tasks, which seriously implies C# too

What makes you say that?

C++ and Rust have options, to name a couple alternatives.

**Reply by DrusTheAxe** [reply]
> when Windows is in sleep mode the periodic task keeps on being fired but the launch of the app seems to get queued. The side effect of this: when Windows goes out of sleep mode suddenly the app is launched 10 to 20 times.

Thats interesting. What version of Windows have you seen this? What kind of activation - UWP app, desktop bridge (aka Centennial), or...? What API are you using to activate the app - ActivateApplication(), Windows.System.Launcher, ShellExecute, CreateProcess on an AppExecutionAlias, other?

Interesting that activation and triggers don't have the same behavior re sleep. Sounds like activation (overtly, or coincidentally something in the call stack) is blocked during sleep.

@brialmsft @jsidewhite @kanismohammed @jonwis do you know more? Is this intentional? A bug?

**Reply by JosHuybrighs** [reply]
@DrusTheAxe 

> What version of Windows have you seen this? What kind of activation - UWP app, desktop bridge (aka Centennial), or...? What API are you using to activate the app - ActivateApplication(), Windows.System.Launcher, ShellExecute, CreateProcess on an AppExecutionAlias, other?

I see it on Windows 11, 22621.2861.
The background task is defined in the package manifest as:

```
<Applications>
  <Application ..>
    <Extensions>
        <Extension Category="windows.backgroundTasks" EntryPoint="SyncFolder.BackgroundTasks.ScheduleBackupTask">
          <BackgroundTasks>
            <Task Type="timer"/>
          </BackgroundTasks>
        </Extension>
   ...
</Applications>
<Extensions>
  <Extension Category="windows.activatableClass.inProcessServer">
    <InProcessServer>
      <Path>SyncFolder\WinRT.Host.dll</Path>
      <ActivatableClass ActivatableClassId="SyncFolder.BackgroundTasks.ScheduleBackupTask" ThreadingModel="both"/>
    </InProcessServer>
  </Extension>
</Extensions>

```
The code of the 'schedular' task is:

```
        public void Run(IBackgroundTaskInstance taskInstance)
        {
            var deferral = taskInstance.GetDeferral();
            string message = "Run";
            try
            {
                LaunchSyncFolder("ScheduleRequest", out message);
            }
            catch (Exception ex)
            {
			    message = ex.Message;
            }
            // Log message
            // ConditionalFileLogger($"SchedularTask - Result: {message}");
            // Removed because ConditionalFileLogger writes to a file in AppData\...\LocalData which doesn't
            // work because getting the folder using Windows.Storage.ApplicationData.Current.LocalFolder gives an exception.
            deferral.Complete();
        }

        void LaunchSyncFolder(string parmId, out string message)
        {
            message = "InitLaunch";
            try
            {
                if (ApiInformation.IsApiContractPresent("Windows.ApplicationModel.FullTrustAppContract", 1, 0))
                {
                    FullTrustProcessLauncher.LaunchFullTrustProcessForCurrentAppAsync(parmId).GetAwaiter().GetResult();
                    message = "Launched";
                }
                else
                {
                    message = "NotFullTrust";
                }
            }
            catch (Exception e)
            {
                message = $"Exception in Launch: {e.Message}";
            }
        }

```

I now see that in order to avoid async in the definition of the Run method I call `FullTrustProcessLauncher.LaunchFullTrustProcessForCurrentAppAsync` with `GetAwaiter().GetResult()` added to it. I asume that this  shouldn't make any difference (unless 'async' in case of a IBackgroundTask has some special meaning). Haven't tested this yet, but I will try to find out.

**Reply by DarranRowe** [reply]
@DrusTheAxe 
> > all in process background tasks need to be migrated to COM background tasks, which seriously implies C# too
> 
> What makes you say that?
> 
> C++ and Rust have options, to name a couple alternatives.

The documentation.

![Screenshot 2023-12-31 134917](https://github.com/microsoft/WindowsAppSDK/assets/52577874/6fbd9142-b6c3-4802-a68d-04173c14d183)

**Reply by JosHuybrighs** [reply]
@DarranRowe @DrusTheAxe 

> One question that I have been meaning to ask. Since [Mapping UWP features to the Windows App SDK](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/migrate-to-windows-app-sdk/feature-mapping-table) has some interesting implications. First, C++ has no such thing as in process background tasks. Secondly, all in process background tasks need to be migrated to COM background tasks, which seriously implies C# too. Have you actually tried using this method for your background tasks?

I looked again at the documentation about porting UWP background tasks to a WinUI/WinAppSDK desktop app and tried the 2 methods proposed there for the 2 type of tasks I have:

- a task that runs every 15 minutes (TimeTrigger - `<Task Type="timer"/>`)
- a task that watches USB device inserts/removals (DeviceWatcherTrigger - `<Task Type="deviceUse"/>`)

**Method 1: COM background tasks**

This requires a COMServer for each task and some provisioning in your app to register at DCOM in order to act as the hosting process for the background task and allow your app to take that role both when opened by the user and not opened. A couple of remarks about this:

- With WinAppSDK 1.4 and VS17.8.3 `RegistrationServices.RegisterTypeForComClients() `is not available in .NET 6,7.. and so you need to do the registration with a p/invoke call to `CoRegisterClassObject()`. You do this each time the app is started.
- To allow your app to be the COMServer when not open you must add for each BG task a `<com:Server..>` definition in the manifest file with an argument that can be checked in the `OnLaunched `method to see if the app is opened by the user or not.
- I got only a `TimeTrigger `BG task working. Registration of the `DeviceWatcherTrigger `task causes the following exception:

```
Exception: System.Runtime.InteropServices.COMException
HResult: 0x80070032
InnerException: null
Message: ""
StackTrace:
at WinRT.ExceptionHelpers.<ThrowExceptionForHR>g__Throw|39_0(Int32 hr)
   at ABI.Windows.ApplicationModel.Background.IBackgroundTaskBuilderMethods.Register(IObjectReference _obj)
   at Windows.ApplicationModel.Background.BackgroundTaskBuilder.Register()
   at SyncFolder.Services.AppBackgroundTasks.<RegisterTasksAsync>d__4.MoveNext() in C:\Users\joshu\Projecten\Eigen\trunk\SyncFolder\SyncFolder\Services\AppBackgroundTasks.cs:line 142

```
Pro: The task has full access to any shared code. It runs with MediumIL I think; only checked this with a Debug build with and without VS.

**Method 2: C# OOP background tasks**

This is the method I have been using from the start. Things to be aware of:

- Unfortunately this method requires that the BG task is defined in a seperate library and is a C#/WinRT component.
- Both tasks (TimeTrigger and DeviceWatcherTrigger) work without problems.
- The challenging part is: no BG task can run without having some kind of access to data that is provisioned and managed by the main app, be it a repository file or appsettings.  How do you access this data?
Being a C#/WinRT component one more or less has to recode all these classes inside the BG task project since referencing a shared library doesn't work.
That comes with its own problems: because of WinRT 'imported' classes must all become 'internal' and 'async Task<>' type of methods must be reworked.

**Conclusion**

Method 1 has the advantage of running inside the app (it is InProc but not like in UWP) with full access to everything. Unfortunately it doesn't seem to support all type of BG triggers.
For the time being I will therefore stick to method 2.

None of the 2 methods are however straightforward and require considerable effort to make it work. A good document that describes this would be highly appreciated.

**Comment by assassin316** [comment]
I see this topic has gone unanswered, but I agree with @JosHuybrighs  we need more detailed docs on OOP background tasks for WinUI 3, it's been a nightmare to work with. Even AI hasn't solved my issues (I've tried all the latest models and none can get my BG task to run correctly). I'm currently set up as a COM server periodic BG task and have been struggling for week. There needs to be more samples/docs on this topic.

**Reply by assassin316** [reply]
I've seen that sample but it's too basic and not realistic. 

The majority of developers will put the background task on a separate project in a multi-project solution.

Ideally there should be a sample that shows that and not stuff everything in one single project. 

The idea would be to show how the main app registers and sets up the background task project. In my case my solution consists of four projects: 

The main app 
The package project (msix)
The background task 
The core (shared by the main app and the background task)
